Target-dependent translation of inline asm

fw-immunant opened the pull request:
This would benefit greatly from feedback on how it's detecting target architecture, which is currently pretty hacky and completely undocumented, and will also need some consideration on how we can test target-architecture-specific functionality in CI (I want to make sure installing toolchains for multiple architectures doesn't slow down CI as a whole too much, and we need to have a way to validate translation of programs for non-host architectures, perhaps without running them).

ahomescu commented on c2rust-transpile/src/lib.rs:
```diff
@@ -218,9 +219,32 @@ fn get_module_name(
     file.to_str().map(String::from)
 }
 
+fn target_arch_from_args(clang_args: &[& str]) -> Option<String> {
```
Clang knows the target for the original source file, so couldn't we export that from `AstExporter` instead?

fw-immunant replied:
Good call, done.

fw-immunant commented:
I've fixed a couple more bugs and added tests which will also test other architectures (just arm for now) as much as their runtime environment permits.

rinon commented on c2rust-transpile/src/translator/assembly.rs:
```diff
@@ -101,7 +131,24 @@ fn parse_constraints(mut constraints: &str) ->
+                } else {
+                    warn!("Did not recognize inline asm constraint: {}\n\
+                    It is likely that this will cause compilation errors or \
+                    incorrect semantics in the translated program; please
```
Do you need the escapes? If so it's missing on this line.

fw-immunant replied:
The escape prevents the leading whitespace on the next line from being part of the string, so it is needed. Will fix.

rinon commented on c2rust-transpile/src/translator/assembly.rs:
```diff
@@ -262,13 +391,15 @@ fn asm_is_att_syntax(asm: &str) -> bool {
-        asm.contains('$')
+        // Guess based on sigils used in AT&T assembly. This would be more
+        // robust if it stripped comments first.
+        asm.contains('$') || asm.contains('%') || asm.contains('(')
```
I'm a bit concerned that '(' is too broad and will end up with false positives.

fw-immunant replied:
'(' is intended to pick out `disp(base,idx,scale)` address calculations in AT&T syntax. Is there anywhere it can appear in Intel syntax? The only places I can think of are comments and macros. It's not hard to remove comments, but parsing macros seems like it's more trouble than we should bother with.

fw-immunant replied:
We now remove comments before searching for these sigils.

rinon reviewed (approved):
LGTM, but maybe Andrei should take another look because I haven't really looked at inline asm in a while.

ahomescu commented on c2rust-transpile/src/translator/assembly.rs:
```diff
@@ -186,7 +314,8 @@ fn reg_is_reserved(constraint: &str) -> Option<(&str, &str)> {
 /// Modifies operands and returns a pair of prefix and suffix strings that
 /// should be appended to the assembly template.
-fn rewrite_reserved_reg_operands(att_syntax: bool, operands: &mut Vec<BidirAsmOperand>) -> (String, String) {
+fn rewrite_reserved_reg_operands(att_syntax: bool, arch: Arch,
```
Run rustfmt? This doesn't look like how it would break this line.

fw-immunant replied:
Done.

ahomescu commented on c2rust-transpile/src/translator/assembly.rs:
```diff
@@ -239,6 +368,35 @@ fn rewrite_reserved_reg_operands(att_syntax: bool, operands: &mut Vec<BidirAsmOp
 }
 
+/// Removes comments from an x86 assembly template.
+fn remove_comments(mut asm: &str) -> String {
```
Could we keep the comments and prepend them to the Rust assembly statement?

fw-immunant replied:
The comment-stripped asm is only used for the "AT&T or Intel" heuristic. I can add a comment to clarify that.

ahomescu commented on tests/asm.arm/Cargo.toml:
```diff
@@ -0,0 +1,6 @@
+[package]
```
Could you also add an arm64 test? I think people might care about that more than 32-bit.

fw-immunant replied:
Will do.

fw-immunant replied:
Done.

fw-immunant commented on c2rust-transpile/src/translator/assembly.rs:
```diff
@@ -117,27 +164,102 @@ fn parse_constraints(mut constraints: &str) ->
+        parsed_constraint.starts_with(|c: char| c.is_ascii_digit())
+}
+
+/// Translate an architecture-specific assembly constraint from llvm/gcc
```
According to https://llvm.org/docs/LangRef.html#constraint-codes, constraint codes generally match between LLVM and GCC, but [GCC's semantics are considered canonical](https://llvm.org/docs/LangRef.html#supported-constraint-code-list).

The assembly template we get has been processed by Clang and c2rust-ast-exporter, and has register references prefixed with dollar signs, so I believe it's LLVM's format. But I don't know where to find docs on that format itself, so I've mostly developed the parsing here by trial-and-error.

ahomescu commented on c2rust-transpile/src/translator/assembly.rs:
```diff
@@ -291,6 +498,12 @@ fn rewrite_asm<F: Fn(&str) -> bool>(asm: &str, is_mem_only: F) -> String {
         }
 
+        // Note empty chunks
+        if chunk == "" {
```
Minor nit: `chunk.is_empty()` could have worked too.
